### PR TITLE
Allow setting up an auto-resubscribe ReadClient with just a peer id.

### LINKED ISSUE #23983: Add a way to initialize a subscription-with-resubscribe ReadClient with just a ScopedNodeId, not a session
Right now, doing a SendAutoResubscribeRequest will handle setting up CASE as needed if the session ever drops, but not for the initial session setup.  That means that if you want retries on the original session setup you have to do it manually.

I was talking to @mrjerryjohns and he suggests that we add a way to have exactly one of `ScopedNodeId` and a session in `ReadPrepareParams`, and then the `ReadClient` could handle the session bits as needed, effectively running the "resubscribe" machinery up front to get the session.  This might simplify the logic for consumers a bit.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
This delegates all the work of setting up the CASE session, and the attendant retries, to the ReadClient.

Fixes https://github.com/project-chip/connectedhomeip/issues/23983

Testing: I did manual testing using a hacked-up version of chip-tool that ensured we had no existing session before calling the new API, and made sure that CASE got established and that if the server was not available at that point we did the normal auto-resubscribe retries of CASE until it did become available.

### CODE REVIEW

File: src/app/ReadClient.h [hunk 1]
```diff
@@ -352,12 +352,23 @@ class ReadClient : public Messaging::ExchangeDelegate
...
 
+    /**
+     * Like SendAutoResubscribeRequest above, but without a session being
+     * available in the ReadPrepareParams.  When this is used, the ReadClient is
+     * responsible for setting up the CASE session itself.
+     *
+     * When using this version of SendAutoResubscribeRequest, the
+     * ReadPrepareParams should have a reference to a session, and if it does
+     * have one that session will be ignored.
+     */
```
Comment (jtung-apple) [inline src/app/ReadClient.h:369]: Would it be simpler to say "When using this version of SendAutoResubscribeRequest, any session to which ReadPrepareParams has a reference will be ignored."?

**Review by jtung-apple — APPROVED** [review]

File: src/app/ReadClient.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/ReadClient.h:369]: Yes, good idea.  Especially since I lost a "not" in my comments!

**Review by woody-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Review by yunhanw-google — COMMENTED** [review]
is that possible to write a test case for this? expire the session for subscription, then call this new API to create the session?

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
> is that possible to write a test case for this? 

I tried to; testing CASESessionManager and anything that uses it is a pain, because it needs to do operational DNS-SD lookups....  I did spend some time this morning on it @yunhanw-google and could not get something sane in place so far; you have to bring up a _lot_ of the stack to test this.

I am going to see whether I can write some end-to-end tests once MTRDevice is using this.

**Comment by yunhanw-google** [comment]
I see, thanks for the reply.

**Pull request merged**